c++/triqs/atom_diag/impl/atom_diag.cpp

```diff
@@ -107,17 +108,43 @@ namespace triqs {
     auto atom_diag<Complex>::get_matrix_element_of_monomial(operators::monomial_t const &op_vec, int B) const
       -> std::pair<int, matrix_t> {
 
-      auto m = nda::eye<scalar_t>(get_subspace_dim(B));
+      imperative_operator<class hilbert_space, scalar_t, false>
+        monomial_op(many_body_op_t(1.0, op_vec), fops);
 
-      auto const &fops = get_fops();
-      for (int i = op_vec.size() - 1; i >= 0; --i) {
-        int ind = fops[op_vec[i].indices];
-        int Bp  = (op_vec[i].dagger ? cdag_connection(ind, B) : c_connection(ind, B));
-        if (Bp == -1) return {-1, std::move(m)};
-        m = (op_vec[i].dagger ? cdag_matrix(ind, B) : c_matrix(ind, B)) * m;
-        B = Bp;
+      matrix_t m;
+      int Bp = -1;
+      bool found_nonzero_element = false;
+
+      for(auto [i_index, i]: itertools::enumerate(get_fock_states(B))) {
+        state<class hilbert_space, scalar_t, true> initial_st(full_hs, i);
+
+        auto final_st = monomial_op(initial_st);
+
+        // The initializer for i_index is needed here because of the Core Language Defect #2313.
+        // https://stackoverflow.com/a/46115028
+        foreach(final_st, [&, i_index = i_index, j_count = 0](fock_state_t j, scalar_t x) mutable {
+          assert(++j_count == 1);
+          for(auto const& sp : sub_hilbert_spaces) {
+            if(sp.has_state(j)) {
+              Bp = sp.get_index();
+
+              if(not found_nonzero_element) {
```

Comment:
We could get rid of `found_nonzero_element` by checking if `m` is empty?
```suggestion
              if(m.empty()) {
```